cmd/snapd-apparmor/main_test.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/snapcore/snapd/testutil"
+)
+
+// Hook up check.v1 into the "go test" runner
+func Test(t *testing.T) { TestingT(t) }
+
+type mainSuite struct {
+	testutil.BaseTest
+}
+
+var _ = Suite(&mainSuite{})
+
+func (s *mainSuite) SetUpTest(c *C) {
+	dirs.SetRootDir(c.MkDir())
+
+}
+
+func (s *mainSuite) TearDownTest(c *C) {
+	dirs.SetRootDir("/")
+}
+
+func (s *mainSuite) TestIsContainerWithInternalPolicy(c *C) {
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	appArmorSecurityFSPath := filepath.Join(dirs.GlobalRootDir, "/sys/kernel/security/apparmor/")
+	err := os.MkdirAll(appArmorSecurityFSPath, 0755)
+	c.Assert(err, IsNil)
+
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	// simulate being inside WSL
+	testutil.MockCommand(c, "systemd-detect-virt", "echo wsl")
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, true)
+
+	// simulate being inside a container environment
+	testutil.MockCommand(c, "systemd-detect-virt", "echo lxc")
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	f, err := os.Create(filepath.Join(appArmorSecurityFSPath, ".ns_stacked"))
+	c.Assert(err, IsNil)
+	f.WriteString("yes")
+	f.Close()
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	f, err = os.Create(filepath.Join(appArmorSecurityFSPath, ".ns_name"))
+	c.Assert(err, IsNil)
+	defer f.Close()
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+
+	f.WriteString("foo")
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, false)
+	// lxc/lxd name should result in a container with internal policy
+	f.Seek(0, 0)
+	f.WriteString("lxc-foo")
+	c.Assert(snapd_apparmor.IsContainerWithInternalPolicy(), Equals, true)
+}
+
+func (s *mainSuite) TestLoadAppArmorProfiles(c *C) {
+	parserCmd := testutil.MockCommand(c, "apparmor_parser", "")
+	defer parserCmd.Restore()
+	err := snapd_apparmor.LoadAppArmorProfiles()
+	c.Assert(err, IsNil)
+	// since no profiles to load the parser should not have been called
+	c.Assert(parserCmd.Calls(), DeepEquals, [][]string(nil))
+
+	// mock a profile
+	err = os.MkdirAll(dirs.SnapAppArmorDir, 0755)
+	c.Assert(err, IsNil)
+
+	profile := filepath.Join(dirs.SnapAppArmorDir, "foo")
+	f, err := os.Create(profile)
+	c.Assert(err, IsNil)
+	f.Close()
+
+	// ensure SNAPD_DEBUG is set in the environment so then --quiet
+	// will *not* be included in the apparmor_parser arguments (since
+	// when these test are run in via CI SNAPD_DEBUG is set)
+	os.Setenv("SNAPD_DEBUG", "1")
+	err = snapd_apparmor.LoadAppArmorProfiles()
+	c.Assert(err, IsNil)
+
+	// check arguments to the parser are as expected
+	c.Assert(parserCmd.Calls(), DeepEquals, [][]string{
+		{"apparmor_parser", "--replace", "--write-cache",
+			"-O", "no-expr-simplify",
+			fmt.Sprintf("--cache-loc=%s/var/cache/apparmor", dirs.GlobalRootDir),
+			profile}})
+
+	// test error case
+	testutil.MockCommand(c, "apparmor_parser", "exit 1")
```

Comment:
(niptick^3) maybe include some error output here so that we can test that error output reporting below also works?

Reply:
That is a good idea - done in 09473d4fa3